Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,545 @@
+    io_uring_register_files_update_fp io_uring_register_files_update;
+    io_uring_register_eventfd_fp io_uring_register_eventfd;
+    io_uring_register_eventfd_async_fp io_uring_register_eventfd_async;
+    io_uring_unregister_eventfd_fp io_uring_unregister_eventfd;
+    io_uring_register_probe_fp io_uring_register_probe;
+    io_uring_register_personality_fp io_uring_register_personality;
+    io_uring_unregister_personality_fp io_uring_unregister_personality;
+    __io_uring_sqring_wait_fp __io_uring_sqring_wait;
+    __io_uring_get_cqe_fp __io_uring_get_cqe;
+} liburing_functions;
+
+// Convenience macro for resolving
+#define _DL_RESOLVE(symbol) \
+    liburing_functions.symbol = (symbol ## _fp) dlsym(dl_handle, #symbol);  \
+    if ((err = dlerror()) != NULL) {  \
+        printf("WARNING: Failed to resolve " #symbol " from liburing, falling back on epoll()\n");  \
+        (void) dlclose(dl_handle); \
+        return -1;  \
+    }
+
+// dynamically load liburing and resolve symbols. Should be called once before using io_uring.
+// returns 0 on successful loading and resolving of functions, otherwise error
+
+// Definitons from syscall.c in liburing, we'll just do a syscall that will fail
+// to check for availability of the interface - better than checking specific kernel
+// versions as support can be backported to earlier kernels
+
+#ifdef __alpha__
+/*
+ * alpha is the only exception, all other architectures
+ * have common numbers for new system calls.
+ */
+# ifndef __NR_io_uring_setup
+#  define __NR_io_uring_setup        535
+# endif
+# ifndef __NR_io_uring_enter
+#  define __NR_io_uring_enter        536
+# endif
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    537
+# endif
+#else /* !__alpha__ */
+# ifndef __NR_io_uring_setup
+#  define __NR_io_uring_setup        425
+# endif
+# ifndef __NR_io_uring_enter
+#  define __NR_io_uring_enter        426
+# endif
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    427
+# endif
+#endif
+
+int _check_syscall_available() {
+    errno = 0;
+    if (syscall(__NR_io_uring_register, 0, IORING_UNREGISTER_BUFFERS, NULL, 0) && errno == ENOSYS) {
+        return -1;
+    }
+    return 0;
+}
+
+int _check_capabilities() {
+    struct io_uring_probe *probe = CNIOLinux_io_uring_get_probe();
+    int capabilities_check;
+    
+    capabilities_check = CNIOLinux_io_uring_opcode_supported(probe, IORING_OP_POLL_ADD) &
+                         CNIOLinux_io_uring_opcode_supported(probe, IORING_OP_POLL_REMOVE);
+
+    CNIOLinux_io_uring_free_probe(probe);
+    return capabilities_check == 1 ? 0 : -1;
+}
+
+#define SWIFTNIO_URING_DEFAULT_SIZE 8192 // needed for malloc tests with oneshot polls, otherwise 4K suffices for most tests.
+#define SWIFTNIO_URING_MIN_SIZE 4
+#define SWIFTNIO_URING_MAX_SIZE 65536
+
+unsigned int CNIOLinux_io_uring_ring_size() {
+    const char *uring_size = getenv("SWIFTNIO_URING_RING_SIZE");
```

Comment:
there's no prior art in SwiftNIO for using environment variables to configure something. I think we should pull this for now and add this functionality in a follow-up. Probably best done using some option on the bootstrap anyway I think.